src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -120,12 +120,6 @@ void OnAnalysisException()
 
         public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
         {
-            // Perf optimization. check whether we want to analyze this project or not.
```

Comment:
This early-out seems incorrect to do here. If nothing else, if you turned full-solution-analysis on, and then got diagnostics in the error list, and turned FSA back off again, we'd immediately hit the fast path here so nothing would ever cause the FSA diagnostics to disappear until you opened each file. This also meant that our analyzer that reports file load issues wouldn't report visible diagnostics unless you had FSA on which meant it wasn't terribly useful either.

There are still checks for FSA being on in GetProjectAnalysisDataAsync; we still rely on that being there and not actually doing full analysis unless needed.

Reply:
(see the PR text for more details)

Reply:
This sounds reasonable, but reasonable things can still be scary! I presume `project.GetSourceGeneratorDiagnosticsAsync` will still be expensive in terms of forcing the compilation, symbols and running all generators?

Reply:
@mavasani: it's expensive in that the first person to do it isn't cheap. But anybody asking for a compilation is paying that same cost; my goal here is we're only going to be asking for compilations that are already open documents or in their dependency chain. In essence, the analysis of open documents will have already ran all of that code -- and once it's asked for once, it's fast for everybody else.